docs/source/custom-collections.rst

```diff
@@ -0,0 +1,581 @@
+             without going through a deprecation cycle.
+
+.. note:: This is considered an advanced feature. For most cases the built-in
+          collections are probably sufficient.
+
+Before reading this you should read and underestand:
+
+- :doc:`overview <graphs>`
+- :doc:`graph specification <spec>`
+- :doc:`custom graphs <custom-graphs>`
+
+**Contents**
+
+- :ref:`Description of the dask collection interface <collection-interface>`
+- :ref:`How this interface is used to implement the core dask
+  methods <core-method-internals>`
+- :ref:`How to add the core methods to your class <adding-methods-to-class>`
+- :ref:`example-dask-collection`
+- :ref:`How to check if something is a dask collection <is-dask-collection>`
+- :ref:`How to make tokenize work with your collection <deterministic-hashing>`
+
+
+.. _collection-interface:
+
+The Dask Collection Interface
+-----------------------------
+
+To create your own dask collection, you need to fullfill the following
+interface. Note that there is no required base class.
+
+It's recommended to also read :ref:`core-method-internals` to see how this
+interface is used inside dask.
+
+
+.. method:: __dask_graph__(self)
+
+    The dask graph.
+
+    Returns
+    -------
+    dsk : MutableMapping, None
+        The dask graph. If ``None``, this instance will not be interpreted as a
+        dask collection, and none of the remaining interface methods will be
+        called.
+
+
+.. method:: __dask_keys__(self)
+
+    The output keys for the dask graph.
+
+    Returns
+    -------
+    keys : list
+        A possibly nested list of keys that represent the outputs of the graph.
+        After computation, the results will be returned in the same layout,
+        with the keys replaced with their corresponding outputs.
+
+
+.. staticmethod:: __dask_optimize__(dsk, keys, \*\*kwargs)
+
+    Given a graph and keys, return a new optimized graph.
+
+    This method can be either a ``staticmethod`` or a ``classmethod``, but not
+    an instancemethod.
+
+    Note that graphs and keys are merged before calling ``__dask_optimize__``;
+    as such the graph and keys passed to this method may represent more than
+    one collection sharing the same optimize method.
+
+    If not implemented, defaults to returning the graph unchanged.
+
+    Parameters
+    ----------
+    dsk : MutableMapping
+        The merged graphs from all collections sharing the same
+        ``__dask_optimize__`` method.
+    keys : list
+        A list of the outputs from ``__dask_keys__`` from all collections
+        sharing the same ``__dask_optimize__`` method.
+    \*\*kwargs
+        Extra keyword arguments forwarded from the call to ``compute`` or
+        ``persist``. Can be used or ignored as needed.
+
+    Returns
+    -------
+    optimized_dsk : MutableMapping
+        The optimized dask graph.
+
+
+.. staticmethod:: __dask_default_get__(dsk, keys, \*\*kwargs)
```

Comment:
We may want to change the name of `get` in the future to something more informative.  Is there a different name that we can use here that would be more future-proof?  `__dask_scheduler__` ?

Reply:
Good idea, done.